rfcs/0155-nixos-migrations.md

```diff
@@ -0,0 +1,118 @@
+
+```Nix
+{
+  config.migrations = [ { version = "2023-04-07"; ... } ];
+}
+```
+
+The version information will be stored as a formatted date string (such that lexigraphical and chronological sort are the same), representing the date at which the migration was merged into Nixpkgs.
+
+## Up Script
+
+A place in the Nix Module system for describing the imperative steps required to migrate from the previous version to the new version. This will be run as a systemd "oneshot" service, to take advantage of the standard architecture.
+
+```Nix
+{
+  up.script = ''
+    ${pkgs.my-service-cli} --run-migration /etc/service-state
+  '';
+}
+```
+
+# Down Script
+
+A place in the Nix Module system for describing the imperative steps required to migrate from the current version to the previous version. This will run as a systemd "oneshot" service, to take advantage of the standard architecture. 
+
+```Nix
+{
+  down.script = "my-service.up" ''
+    ${pkgs.my-service-cli} --run-migration /etc/service-state
+  '';
+}
+```
+
+## Transactionality and Warnings
+
+There are limitations for to the effectiveness of up and down scripts. For example, some migrations are destructive in nature, meaning returning to a previous version is not possible. Sometimes a service cannot be upgraded. Human error and entropy can cause these scripts to fail to execute. In a database migration context we can rely on transactions to cope with some of these issues, a luxury we do not have in this context. As such the onus is placed on the module author to provide these properties to the best of their ability. 
+
+### Known destructive migration
+
+```Nix
+{
+  up.warn = ''
+    The 2023 version of My Service requires the deletion of configuration files. Therefore downgrading in an automated fashion is not possible. This is due to unfortunate hard coding of paths in My Service. You can find your old configuration files retained at /etc/service-state-old for your records. 
+  '';
+  down.warn = ''
+    Downgrade was not attempted as the 2023 upgrade is known to be destructive. If you retained your old configuration files, you need to copy them back to /etc/service-state manually.
+  '';
+}
+```
+
+## Failure potential
+
+```Nix
+{
+  up.backup = ''
+    # backup stateful directory
+    cp -r /etc/my-state $BACKUP
+  '';
+  up.restore = ''
+    # restore stateful directory as the script failed
+    cp -r $BACKUP /etc/my-state
+  '';
+}
+```
+
+`$BACKUP` is a path to a temporary filesystem location which will be deleted upon completion of the migration. This provides a temporary location to backup state if needed. `up.backup` is a hook that will run before `up.script`. If `up.script` encounters an error, `up.restore` is run to ensure that the failed migration does not result in contamination of the system. These options are available in both `up` and `down` definitions. This roughly allows for transaction-like logic for the migration.
```

Comment:
> `$BACKUP` is a path to a temporary filesystem location

I really hope you don't mean it's on `tmpfs`. I think a different wording should be used here.

Regardless, I don't think it's a good idea to clear backups as part of a script, even if the script ran successfully. There might be an error in the logic for all you know. Lastly, we REALLY need this to be as transactional as possible, though thankfully there's a lot of know-how in the Nix space around that.